### PR TITLE
PanelChrome: Menu is wrapped in a render prop for full outside control

### LINKED ISSUE #59085: PanelChrome: Menu is wrapped in a render prop for full outside control
Part of epic https://github.com/grafana/grafana/issues/56163.

The current implementation of passing and rendering the menu inside the PanelChrome is very static.

- [x] Implement the menu similar to the implementation here: [PanelHeader with MenuWrapper](https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/dashgrid/PanelHeader/PanelHeader.tsx#L67)

Necessary changes discussed in PR: https://github.com/grafana/grafana/pull/57203#discussion_r1026154152
> We need this to be behind a render prop (render callback), it is expensive to know all the menu actions (might require loading the panel plugin etc, figuring out actions and permissions, might need to check current data state state).
> 
> We can do all that when the menu opens instead if we put this behind a render function. I think that is how the current menu works (only evaluated when you open it).
> 
> The Dropdown component the overlay prop takes a function as well, and it is only called when the overlay is opened
@torkelo 

**N.B.** Do not implement menu in absolutely positioned box because of problems of the sort: https://github.com/grafana/grafana/issues/25930. Best is to use the `Dropdown` because it wraps the `overlay` in a `Portal`.

### PR DESCRIPTION
**Pull request by polibb**
**What is this feature?**
Allows panels rendered with `PanelChrome` (`grafana-ui`) to have a **menu**.
The `Dropdown` (`grafana-ui`) handles opening/closing of the menu. It uses a [Portal](https://reactjs.org/docs/portals.html) to render the menu. 

**Why do we need this feature?**
We need it in the dashboards  when panels are implemented using `PanelChrome` 😊.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/59085

**Special notes for your reviewer**:
To do:
- [x] Menu icon is on the right in a fixed header.
- [x] Streaming can be stopped from menu
- [x] Add Storybook examples of having a menu :). And not having a menu.

### CODE REVIEW

**Review by dprokop — APPROVED** [review]
@polibb - this looks good code-wise. Can I test it somehow? 🐒 

In the future, please avoid mixing API changes with refactors (in this case `const` to `function`) as those are technically irrelevant changes to the scope of the PR and are blurring the picture a bit.

**Comment by polibb** [comment]
> @polibb - this looks good code-wise. Can I test it somehow? 🐒
> 
> In the future, please avoid mixing API changes with refactors (in this case `const` to `function`) as those are technically irrelevant changes to the scope of the PR and are blurring the picture a bit.

Yes, use the feature flag `newPanelChromeUI` and open any dashboard 🙏. 
Because the `menu` prop is passed down by PanelStateWrapper the (almost) same way it is implemented right now, it should handle all the same functionalities we can get from the panel menu.

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuProvider.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ interface Props {
   children: (props: PanelHeaderMenuProviderApi) => ReactElement;
 }
 
-export const PanelHeaderMenuProvider: FC<Props> = ({ panel, dashboard, children }) => {
+export function PanelHeaderMenuProvider({ panel, dashboard, children }: Props) {
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuProvider.tsx:20]: This provider component look like it can just be rewritten to a hook does not need to be a component using children render prop

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuWrapper.tsx [hunk 2]
```diff
@@ -11,12 +11,12 @@ interface Props {
   onClose: () => void;
 }
 
-export const PanelHeaderMenuWrapper: FC<Props> = ({ panel, dashboard }) => {
+export function PanelHeaderMenuWrapper({ panel, dashboard }: Props) {
   return (
     <PanelHeaderMenuProvider panel={panel} dashboard={dashboard}>
       {({ items }) => {
         return <PanelHeaderMenu items={items} />;
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuWrapper.tsx:18]: My hope / plan was to not use this component as it is using some very old menu styles / css classes, and is not using our Menu component from grafana/ui 

But we can do that switch in another PR

**Comment by torkelo** [comment]
Took this for a spin, and found some minor issues

* The menu icon is in the middle of the header and not in the far right corner
* The placement is too far down (should be just below the menu icon) 

Screenshot 2022-12-21 at 16 12 58

**Review by kaydelaney — APPROVED** [review]
Seems good in terms of code!

File: packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx [hunk 3]
```diff
@@ -100,15 +100,15 @@ export function PanelChrome({
   const showStreaming = loadingState === LoadingState.Streaming && !isUsingDeprecatedLeftItems;
 
   const renderStatus = () => {
-    if (isUsingDeprecatedLeftItems) {
-      return <div className={cx(styles.rightAligned, styles.items)}>{itemsRenderer(leftItems, (item) => item)}</div>;
-    } else {
-      const showError = loadingState === LoadingState.Error || status?.message;
-      return showError ? (
+    const showError = loadingState === LoadingState.Error || status?.message;
+    if (!isUsingDeprecatedLeftItems && showError) {
```
Comment (torkelo) [inline packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:104]: Don't think this !leftItems check here is needed, loadingState and status are new props so should this should not show for any old uses

File: public/app/features/dashboard/dashgrid/PanelStateWrapper.tsx [hunk 4]
```diff
@@ -589,13 +590,23 @@ export class PanelStateWrapper extends PureComponent<Props, State> {
     const title = panel.getDisplayTitle();
     const padding: PanelPadding = plugin.noPadding ? 'none' : 'md';
 
+    let menu;
+    if (!dashboard.meta.publicDashboardAccessToken) {
+      menu = (): React.ReactElement => (
+        <div data-testid="panel-dropdown">
+          <PanelHeaderMenuWrapper panel={panel} dashboard={dashboard} onClose={() => {}} />
```
Comment (torkelo) [inline public/app/features/dashboard/dashgrid/PanelStateWrapper.tsx:597]: Can we add a remaining todo do use the menu component from grafana/ui instead (in a follow up PR)

Comment (polibb) [reply public/app/features/dashboard/dashgrid/PanelStateWrapper.tsx:597]: Yes 🙌

File: packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx (hunk 3 above)
Comment (polibb) [reply packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:104]: Yes, I have this in my to-do for a separate PR so as to not put irrelevant things in here 🙌.
We added it at first because the leftItems and new state positions were both on the far right of the header, so could not exist together.

Comment (polibb) [reply packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:104]: I moved the render of `leftItems` out of this `renderStatus()` function because they are on the right side of the header together with the `menu`, which is much easier if they are wrapped together with `margin-left: auto` under the _flex_ container that is the header.

File: public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuProvider.tsx (hunk 1 above)
Comment (polibb) [reply public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderMenuProvider.tsx]: Yes, let's do it in a different PR

File: public/app/features/dashboard/utils/getPanelMenu.test.ts [hunk 5]
```diff
@@ -94,117 +95,285 @@ describe('getPanelMenu', () => {
...
-          },
-        ]
-      `);
-    });
+  it('should return the correct panel menu items when data is streaming', () => {
+    const panel = new PanelModel({});
+    const dashboard = createDashboardModelFixture({});
+
+    const menuItems = getPanelMenu(dashboard, panel, LoadingState.Streaming);
+    expect(menuItems).toMatchInlineSnapshot(`
```
Comment (torkelo) [inline public/app/features/dashboard/utils/getPanelMenu.test.ts:103]: if streaming is the thing you want to test here, better to check that items exists rather than big snapshot. This can fail for many reasons not related to loading state (IE use snapshot tests sparingly)

**Review by torkelo — APPROVED** [review]

File: public/app/features/dashboard/utils/getPanelMenu.test.ts (hunk 5 above)
Comment (polibb) [reply public/app/features/dashboard/utils/getPanelMenu.test.ts]: Yes, snapshots are expensive.. good catch, thanks! Fixed 🙏
